### ISSUE TITLE
What should be the coordinates for the GDP ragged arrays and others?

### ISSUE THREAD

**Issue by selipot** [Cloud-Drift/clouddrift-examples#18]
We've been thinking about dimensions and coordinates for the RaggedArray class (and their xarray forms). The xarray documentation [states](https://xarray-test.readthedocs.io/en/latest/data-structures.html):

_xarray uses dims and coords to enable its core metadata aware operations. Dimensions provide names that xarray uses instead of the axis argument found in many numpy functions. Coordinates enable fast label based indexing and alignment, building on the functionality of the index found on a pandas [DataFrame](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame) or [Series](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html#pandas.Series)._

How does this apply to Lagrangian datasets? 
- For the GDP organized as a ragged array I tend to think that there are only two dimensions (obs and traj), that time is perhaps a coordinate (because being regular it can be shared across several drifters), and that everything else is data (lat and lon values are never shared based on their precision). 
- For synthetic Lagrangian trajectories like the ones generated from Parcels, it is maybe a bit different. For experiment for which particles are drifting at the same time, obs and traj are still dimensions and time is a coordinate when the data are organized as 2D arrays. When they are organized as ragged arrays, time may or may not be a coordinates.

Thoughts?

**Comment by milancurcic** [comment]
Here's a snippet of our current xarray GDP representation:

```
<xarray.Dataset>
Dimensions:                (traj: 100, obs: 939619)
Coordinates:
    time                   (obs) float64 9.793e+08 9.793e+08 ... 1.586e+09
    lon                    (obs) float32 71.29 71.28 71.29 ... -152.4 -152.4
    lat                    (obs) float32 -48.15 -48.14 -48.13 ... 21.72 21.72
    ids                    (obs) int64 6428 6428 6428 ... 68246720 68246720
Dimensions without coordinates: traj, obs
Data variables: (12/55)
    ID                     (traj) int64 6428 13566 17927 ... 67772920 68246720
    rowsize                (traj) int32 19132 6631 17906 5738 ... 3592 221 1843
    WMO                    (traj) int32 1600536 3400513 ... 5102772 4601712
    expno                  (traj) int32 9435 7325 9435 ... 21312 21312 21312
    deploy_date            (traj) float32 9.792e+08 1.015e+09 ... 1.585e+09 nan
    deploy_lat             (traj) float64 -48.19 -58.01 -47.04 ... -3.0 24.27
    ...                     ...
```

I don't know the full rationale for the current choice of dims, coords, and data, but I think it should be:

* Dimensions: `traj` (or `ID`, `ids`, pick one), `obs` (or `time`, pick one)
* Coordinates: `traj` (or `ID`, `ids`, pick one), `obs` (or `time`, pick one)
* Data: everything else

In other words, the changes to our current representation should be:

1. `traj` and `obs` are coordinates and dimensions
2. Move `lon` and `lat` to data
3. Remove `ids` and `ID` and just use `traj`.

> For the GDP organized as a ragged array I tend to think that there are only two dimensions (obs and traj), that time is perhaps a coordinate (because being regular it can be shared across several drifters),

@selipot this confused me. Is time different from obs and how? You mention time being regular, which makes me think that maybe here you mean the monotonic and continuous time that ranges from the start time of the earliest trajectory to the end time of the latest trajectory in the dataset? If yes, we would need a new coordinate for that. But what would be the purpose of this time?

**Comment by selipot** [comment]
- Dimensions are scalars. For Lagrangian data I see only two dimensions: `traj` and `obs`.

- Coordinates are vectors (1-D arrays) of distinct values, generally sorted. 

For Lagrangian data organized in 2D arrays (as output by ocean parcels as an example), **`obs`** and **`traj`** are coordinates with names equal to the dimensions. A 30-day simulation with hourly output will have an **`obs`** coordinate with values `range(0,720)`. The **`traj`** coordinate will be a vector of sorted numbers.

For Lagrangian data organized in ragged arrays, I actually do not envision any "physical" coordinate. For the GDP cases, the vector `time` certainly has repeated values and is not sorted, the vectors `lat` and `lon` may contain distinct values but are not sorted and not coordinates of anything, the vector `ids` has repeated values and is not sorted. 

In conclusion I thing Lagrangian ragged arrays have no coordinate at all. Another distinction from what xarray would expect. Unless we define some arbitrary ones for computational reasons? They could be **`traj`** `= range(0,traj)` and **`obs`** `=range(0,obs)`. Would this be useful?

In conclusion, I suggest for the GDP ragged arrays
1. `traj` and `obs` are dimensions
2. There are no coordinate, or we make some up as above if useful for computations/analysis.
3. `time`, `lat`, `lon`, `lon360`, `ids` are all data variables
4. We keep both `ID` with dimension `traj` and `ids` with dimension `obs` for practical reasons for now.

**Comment by milancurcic** [comment]
> Coordinates are vectors (1-D arrays) of distinct values, generally sorted.

Is this true? Our current ragged GDP xarray dataset has coordinates with values that are not distinct nor sorted. So it seems possible (and IMO useful) to have such coordinates.

> There are no coordinate, or we make some up as above if useful for computations/analysis.

If we want xarray, we have to. It seems to me more useful to ask what our data structure should look like (shape, order of elements, etc.) and the coordinates will turn up on their own.

**Comment by selipot** [comment]
> > Coordinates are vectors (1-D arrays) of distinct values, generally sorted.
> 
> Is this true? Our current ragged GDP xarray dataset has coordinates with values that are not distinct nor sorted. So it seems possible (and IMO useful) to have such coordinates.

Our current GDP ragged arrays opened as a xarray dataset are the way they are because we made that up, i.e. that is what is coded in the RaggedArray class. I am obviously questioning this again. I am yet to see a use for these coordinates. When saved as NetCDF files there are no coordinates, only dimensions.

> > There are no coordinate, or we make some up as above if useful for computations/analysis.
> 
> If we want xarray, we have to. It seems to me more useful to ask what our data structure should look like (shape, order of elements, etc.) and the coordinates will turn up on their own.

That's why I am asking. As of right now using the time, lat, lon, ids as coordinates do not make sense to me but I would love to be shown otherwise.

**Comment by selipot** [comment]
And I don't think the RaggedArray class/method allows flexibility for all that. This needs to be modified.

**Comment by milancurcic** [comment]
OK, I **think** I understand why I'm confused now. We're using different terminologies. In xarray terminology a coordinate is what we generally call a dimension (e.g. an xarray coordinate maps to a NetCDF dimension). But an xarray dimension is just a label, it's not the same thing as a dimension in math or NetCDF, etc.

It seems to me that only `traj` and `obs` need to be dimensions (in a general or NetCDF meaning of the word) for a ragged array that is stored as a traditional array structure. Everything else is data defined in one of those dimensions.

**Comment by selipot** [comment]
Haha now I am confused by what you think: here is again what xarray states:

_xarray uses dims and coords to enable its core metadata aware operations. Dimensions provide names that xarray uses instead of the axis argument found in many numpy functions. Coordinates enable fast label based indexing and alignment, building on the functionality of the index found on a pandas [DataFrame](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.html#pandas.DataFrame) or [Series](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.html#pandas.Series)._

I think we're going in circles. But we seem to agree. Ragged arrays have no coordinate in the sense of xarray.

**Comment by milancurcic** [comment]
Yes, this will be easier on a call. :)

**Comment by philippemiron** [comment]
A part of this might be created by my fault. But when I coded the `dataformat` module, to get the variables from xarray I was iterating (see [from_xarray](https://github.com/Cloud-Drift/clouddrift/blob/1217c43282e8b4000273f1a6c0789085e71ac27e/clouddrift/dataformat.py#L122)):
- for var in ds.coords.keys()
- for var in ds.data_vars.keys()

So I split the variables into coordinates and basically "others". Right now, it's not at all intuitive to transfer a variable to a coordinate (or vice verse). It might be better to redo this sections and simply interates using:

- for var in ds.variables.keys()

which contains both the coordinates and the variables of a `xr.Dataset` and then inside of this loop when put the variables into coordinates or data as a function of the arguments provided by the user.

**Comment by milancurcic** [comment]
I think you did a pretty good job with the initial design. It's not that consequential whether a data variable is also a coordinate or not. It's a minor detail IMO, but overall I agree with the proposed direction forward.

**Comment by philippemiron** [comment]
Yes, I can try this.

**Comment by philippemiron** [comment]
PR https://github.com/Cloud-Drift/clouddrift/pull/92.
